pkg/workqueue/gcs/gcs.go

```diff
@@ -373,6 +404,58 @@ func (o *inProgressKey) Complete(ctx context.Context) error {
 		"revision_name": env.KnativeRevisionName,
 	}).Observe(time.Now().UTC().Sub(o.attrs.Created).Seconds())
 
+	// Best-effort delete of the dead-letter object, if it exists.
+	deadLetterKey := o.deadLetterKey()
+	if err := o.client.Object(deadLetterKey).Delete(ctx); err != nil {
+		if !errors.Is(err, storage.ErrObjectNotExist) {
+			clog.WarnContextf(ctx, "Failed to delete dead-letter object %q: %v", deadLetterKey, err)
+		}
+	}
+
+	return o.client.Object(o.attrs.Name).Delete(ctx)
+}
+
+// Deadletter implements workqueue.OwnedInProgressKey.
+func (o *inProgressKey) Deadletter(ctx context.Context) error {
+	if o.ownerCancel != nil {
+		o.ownerCancel()
+	}
+	o.rw.RLock()
+	defer o.rw.RUnlock()
+
+	key := strings.TrimPrefix(o.attrs.Name, inProgressPrefix)
+	deadLetterKey := o.deadLetterKey()
+
+	clog.InfoContextf(ctx, "Moving key %q to dead letter queue as %q", key, deadLetterKey)
+
+	// Copy the in-progress task to the dead letter queue
+	copier := o.client.Object(deadLetterKey).CopierFrom(o.client.Object(o.attrs.Name))
+
+	// Preserve metadata
+	copier.Metadata = o.attrs.Metadata
+	if copier.Metadata == nil {
+		copier.Metadata = make(map[string]string)
+	}
+
+	// Clear the lease expiration when copying the object
+	delete(copier.Metadata, expirationMetadataKey)
+
+	// Add metadata about when the key was dead-lettered
+	copier.Metadata[failedTimeMetadataKey] = time.Now().UTC().Format(time.RFC3339)
+
+	// Create the dead letter entry
+	_, err := copier.Run(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to create dead letter entry: %w", err)
+	}
+
+	// Increment the metric for dead-lettered keys
+	mDeadLetteredKeys.With(prometheus.Labels{
+		"service_name":  env.KnativeServiceName,
+		"revision_name": env.KnativeRevisionName,
+	}).Inc()
```

Comment:
I think we want to count the number of keys with this prefix when we're scanning the bucket and `Set` the total cound 🤔

Reply:
Done